src/openrct2-ui/input/InputManager.cpp

```diff
@@ -0,0 +1,385 @@
+            ie.DeviceKind = InputDeviceKind::JoyButton;
+            ie.Modifiers = SDL_GetModState();
+            ie.Button = e.jbutton.button;
+            ie.State = InputEventState::Release;
+            QueueInputEvent(std::move(ie));
+            break;
+        }
+    }
+}
+
+void InputManager::QueueInputEvent(InputEvent&& e)
+{
+    _events.push(e);
+}
+
+void InputManager::CheckJoysticks()
+{
+    constexpr uint32_t CHECK_INTERVAL_MS = 5000;
+
+    auto tick = SDL_GetTicks();
+    if (tick > _lastJoystickCheck + CHECK_INTERVAL_MS)
+    {
+        _lastJoystickCheck = tick;
+
+        _joysticks.clear();
+        auto numJoysticks = SDL_NumJoysticks();
+        for (auto i = 0; i < numJoysticks; i++)
+        {
+            auto joystick = SDL_JoystickOpen(i);
+            if (joystick != nullptr)
+            {
+                _joysticks.push_back(joystick);
+            }
+        }
+    }
+}
+
+void InputManager::Process()
+{
+    CheckJoysticks();
+    HandleModifiers();
+    ProcessEvents();
+    ProcessHoldEvents();
+    HandleViewScrolling();
+}
+
+void InputManager::HandleViewScrolling()
+{
+    if (gScreenFlags & SCREEN_FLAGS_TITLE_DEMO)
+        return;
+
+    auto& console = GetInGameConsole();
+    if (console.IsOpen())
+        return;
+
+    // Shortcut scrolling
+    auto mainWindow = window_get_main();
+    if (mainWindow != nullptr && (_viewScroll.x != 0 || _viewScroll.y != 0))
+    {
+        window_unfollow_sprite(mainWindow);
+    }
+    InputScrollViewport(_viewScroll);
+
+    // Mouse edge scrolling
+    if (gConfigGeneral.edge_scrolling)
+    {
+        if (input_get_state() != InputState::Normal)
+            return;
+
+        if (gInputPlaceObjectModifier & (PLACE_OBJECT_MODIFIER_SHIFT_Z | PLACE_OBJECT_MODIFIER_COPY_Z))
+            return;
+
+        GameHandleEdgeScroll();
+    }
+}
+
+void InputManager::HandleModifiers()
+{
+    auto modifiers = SDL_GetModState();
+    gInputPlaceObjectModifier = PLACE_OBJECT_MODIFIER_NONE;
+    if (modifiers & KMOD_SHIFT)
+    {
+        gInputPlaceObjectModifier |= PLACE_OBJECT_MODIFIER_SHIFT_Z;
+    }
+    if (modifiers & KMOD_CTRL)
+    {
+        gInputPlaceObjectModifier |= PLACE_OBJECT_MODIFIER_COPY_Z;
+    }
+    if (modifiers & KMOD_ALT)
+    {
+        gInputPlaceObjectModifier |= 4;
+    }
+#ifdef __MACOSX__
+    if (modifiers & KMOD_GUI)
+    {
+        gInputPlaceObjectModifier |= 8;
+    }
+#endif
+
+    if (gConfigGeneral.virtual_floor_style != VirtualFloorStyles::Off)
+    {
+        if (gInputPlaceObjectModifier & (PLACE_OBJECT_MODIFIER_COPY_Z | PLACE_OBJECT_MODIFIER_SHIFT_Z))
+            virtual_floor_enable();
+        else
+            virtual_floor_disable();
+    }
+}
+
+void InputManager::ProcessEvents()
+{
+    while (!_events.empty())
+    {
+        const auto& e = _events.front();
+        Process(e);
+        _events.pop();
+    }
+}
+
+void InputManager::Process(const InputEvent& e)
+{
+    auto& shortcutManager = GetShortcutManager();
```

Comment:
In my opinion the InputManager should only gather the state and let the other systems such as Chat and Console decide how to handle the input state, not the input manager. I'm thinking of either events or polling, so either
```
auto mouseDown = inputSystem->IsMouseDown(MOUSE_MIDDLE);
```
or
```
inputSystem->addHandler(inputState, []() { /* Callback */ });
```
Systems should be connected and not part of each other. If something else needs input handling it would just grow in code and complexity here.

Reply:
You still need this Process method for your second example then to call the callbacks. Right now, something needs to call the keyboard shortcut handler so that is why I have this. I am happy for you to change it to however you think it should be done afterwards.

Reply:
Yes, all systems require an Update/Process method. Also fair enough.